src/eval/Evaluator.py

```diff
@@ -0,0 +1,160 @@
+        """
+        Reads through the analogy question file.
+        If every word in the analogy line is in the vocabulary
+        we store this line in index form.
+
+        builds:
+            analogy_questions: a [n, 4] numpy array containing
+                               the analogy question's word ids.
+                               where n is the number of
+                               questions not skipped from the
+                               analogy question file.
+        """
+        questions = []
+        questions_skipped = 0
+        with open(self.eval_path, "rb") as analogy_f:
+            for line in analogy_f:
+                if line.startswith(b":"):
+                    continue
+                words = line.strip().lower().split(b" ")
+                ids = [self.word2index.get(w.strip().decode(self.encoding))
+                       for w in words]
+                if None in ids or len(ids) != 4:
+                    questions_skipped += 1
+                else:
+                    questions.append(np.array(ids))
+        self.valid_questions = len(questions)
+        total_questions = self.valid_questions + questions_skipped
+        self.precision = self.valid_questions / total_questions
+        self.analogy_questions = np.array(questions, dtype=np.int32)
+
+    def _build_top_k(self):
+        """
+        Uses a tensorflow graph to build the top k closest words
+        to the vector resulting from (c + b - a).
+        Remember, each valid line of the analogy file is of the form
+
+        a b c d
+
+        And we are trying to check if d is the closest word
+        to (c + b - a) using the cosine similarity.
+
+        Since the indices of words a and b appear in the top_k matrix
+        we need to take a matrix of size k+2.
+        """
+        graph = tf.Graph()
+        with graph.as_default():
+            normalized_emb = tf.nn.l2_normalize(self.embeddings, 1)
+            a_emb = tf.gather(normalized_emb,
+                              self.analogy_questions.T[0])
+            b_emb = tf.gather(normalized_emb,
+                              self.analogy_questions.T[1])
+            c_emb = tf.gather(normalized_emb,
+                              self.analogy_questions.T[2])
+
+            target = c_emb + (b_emb - a_emb)
+
+            dist = tf.matmul(target,
+                             normalized_emb,
+                             transpose_b=True)
+            _, pred_idx = tf.nn.top_k(dist, self.top_results + 2)
+        with tf.Session(graph=graph) as sess:
+            self.top_k = sess.run(pred_idx)
+
+    def _build_clean_top_k(self):
+        """
+        Cleaning the top k matrix. Here we remove all the indices
+        form the words b and c from the top k matrix
+        """
+        clean = []
+        for i, line in enumerate(self.top_k):
+            b_c = list(self.analogy_questions[i][1:3])
+            new = []
+            for idx in line:
```

Comment:
I did not change this part because although in the majority of cases the words b and c have the biggest values, this is not always the case. Using the portuguese corpus we have examples like:

- **analogy** = escrever escrevem trabalhou trabalham
- **top 12 results** = ['escrevem', 'estudavam', 'atuam', 'estrearam', 'regeu', 'bud', 'trabalhou', 'trabalham', 'trabalharam', 'moravam', 'milla', 'atuaram']

Word b ('escrevem') has the biggest value, but word c ('trabalhou') has not the second biggest value.